app/safecast/tasks/device_story/update_metadata.rb

```diff
@@ -0,0 +1,66 @@
+# frozen_string_literal: true
+
+module Tasks
+  module DeviceStory
+    class UpdateMetadata
+      class << self
+        def call
+          new.call
+        end
+      end
+
+      def call
+        each_metadata(parse(fetch(devices_uri))) do |device_urn, metadata|
+          begin
+            find_or_update_device(device_urn, metadata)
+          rescue ActiveRecord::RecordInvalid
+            warn "Could not update device metadata for #{device_urn}."
+          end
+        end
+      end
+
+      private
+
+      def find_or_update_device(device_urn, metadata)
+        device = ::DeviceStory::Device.find_by(device_urn: device_urn)
+        if device
+          device.update!(metadata: metadata)
+        else
+          ::DeviceStory::Device.create!(device_urn: device_urn, metadata: metadata)
+        end
+      end
+
+      def each_metadata(metadatas)
+        metadatas.each do |metadata|
+          device_urn = metadata['device_urn']
+          unless device_urn
+            warn 'Could not find device_urn.'
+            next
+          end
+          yield device_urn, metadata
+        end
+      end
+
+      def parse(json)
+        JSON.parse(json)
+      rescue Psych::SyntaxError
+        warn 'Could not parse device metadata.'
+        []
+      end
+
+      def fetch(uri)
+        res = Net::HTTP.get_response(uri)
+        if res.is_a?(Net::HTTPOK)
+          res.body
+        else
+          warn 'Could not fetch device metadata from ttserve.'
+          '{}'
```

Comment:
Similarly, if there is an error fetching the metadata, shouldn't this raise an exception instead of silently continuing?

Reply:
In this case, it would do nothing, just outputs a warning. `JSON.parse(`{}')` returns an empty `Hash`. `Hash` includes `Enumerable` which responds to `#each`. 

This `#fetch` method would raise exception in case of exceptional cases such as network is not available or so. ttserve is simple HTTP server. It usually returns only `200 OK`. I assume that there is an ALB in front of ttserve. If so, ALB sometimes returns `503 Bad gateway`.  In case of getting 503, I think that it would be better to continue silently.

Reply:
Could you please explain why it would be better to continue silently? There is no data to process, only an error. Shouldn't we either retry the request or exit?

Reply:
Per recording today 👍 on @auspicacious 's call for raising rather than warning and creating an empty object (on non-200) or list (on non-json).

I presume we'll be calling this on something like a 15m schedule, so it's okay to just let it explode and we'll try again in the next run.